### PR TITLE
[Indexer] Add token activity and token claim/listing support

### PR DESCRIPTION
**Pull request by bowenyang007**
### Added tables
* token_activities - this is how you would get the user’s historical token activities
* current_token_pending_claims - tracks current pending offers (if a token is claimed its amount will be 0)
* current_token_listings*, token_escrows*, current_token_escrows* - tracks current buy/sell. Escrow tracks tokens that are "locked"
* Added collection_data_id_hash to all token tables (token data, tokens, etc.) to be able to easily query for the collection. This is a note for the wallet team, please use token_data_id_hash and collection_data_id_hash to query for the token and collection instead of the names directly (which might not work b/c we truncate all the names now). To get the hash, just make a query without the hash (e.g. query for all tokens owned by account X and use the hash to get a specific token_data/collection)

*token listings/purchase/escrow stuff is very rough on the edges so use carefully but we will most likely modify

### Test Plan
Ran locally.
Deployed to testnet and devnet. 
Got hasura loaded with the new tables. 

Sample queries: 
*Current tokens owned by user (in TokenStore)*
```
{
  current_token_ownerships(where: {owner_address: {_eq: "0x21d5fe032affa1c8b10d343e9ad5a5618bc13baf5ed4a674fafaa12c54f416cc"}, table_type:{_eq:"0x1::token::TokenStore"},amount: {_gt: 0}}, order_by: {last_transaction_version: desc}) {
    collection_name
    creator_address
    name
    property_version
    amount
  }
}
```
*current offers from offerer wallet*
```
current_token_pending_claims(where: {from_address: {_eq: "0x6855f822bcec1e886f4a0287411edbe17a903ab3212aeabb0c8643ea16586954"}})
```
*current offers from receiver wallet*
```
current_token_pending_claims(where: {to_address: {_eq: "0x2f8da3ee4008e78a76245eca325ef83137a7ca50123ca78eadfc7a7647bf4d04"}})
```
*tokens sent to user*
```
token_activities(where: {to_address: {_eq: "0x2f8da3ee4008e78a76245eca325ef83137a7ca50123ca78eadfc7a7647bf4d04"}})
```

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4324)

### CODE REVIEW

**Review by CapCap — APPROVED** [review]

File: crates/indexer/migrations/2022-09-22-185845_token_offers/up.sql [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+ADD COLUMN collection_data_id_hash VARCHAR(64) NOT NULL DEFAULT '',
+  ADD COLUMN table_type TEXT NOT NULL DEFAULT '';
+ALTER TABLE current_token_datas
+ADD COLUMN collection_data_id_hash VARCHAR(64) NOT NULL DEFAULT '';
+ALTER TABLE token_datas
+ADD COLUMN collection_data_id_hash VARCHAR(64) NOT NULL DEFAULT '';
+ALTER TABLE tokens
+ADD COLUMN collection_data_id_hash VARCHAR(64) NOT NULL DEFAULT '';
+ALTER TABLE token_ownerships
+ADD COLUMN collection_data_id_hash VARCHAR(64) NOT NULL DEFAULT '';
```
Comment (CapCap) [inline crates/indexer/migrations/2022-09-22-185845_token_offers/up.sql:12]: nullable pls

File: crates/indexer/src/models/token_models/token_bcs_utils.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+/// Convert the vector<u8> that is directly generated from b"xxx"
+pub fn convert_hex(val: String) -> Option<String> {
+    let decoded = hex::decode(val.strip_prefix("0x").unwrap_or(&*val)).ok()?;
+    String::from_utf8(decoded).ok()
+}
+
+/// Convert the json serialized PropertyMap's inner BCS fields to their original value in string format
+pub fn convert_bcs_propertymap(s: Value) -> Option<String> {
```
Comment (CapCap) [inline crates/indexer/src/models/token_models/token_bcs_utils.rs:33]: `property_map`?

File: crates/indexer/src/models/token_models/token_bcs_utils.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                        {
+                            "key": "Yuri",
+                            "value": {"type": "0x1::string::String", "value": "0568656c6c6f"}
+                        },
+                        {
+                            "key": "Sergi",
+                            "value": {"type": "u8", "value": "0a"}
+                        }
+                    ]
+                }
```
Comment (CapCap) [inline crates/indexer/src/models/token_models/token_bcs_utils.rs:116]: ah yes, the russians?

File: crates/indexer/migrations/2022-09-22-185845_token_offers/up.sql (hunk 1 above)
Comment (bowenyang007) [reply crates/indexer/migrations/2022-09-22-185845_token_offers/up.sql:12]: I'll remove the default since the backfill is already done.

**Comment by areshand** [comment]
FYI: we will deprecate the token coin swap before mainnet depending on the GK result today.

**Review by zihan-aptos — APPROVED** [review]

**Comment by bowenyang007** [comment]
@areshand Once we deprecate I'll just remove it. I'm not super worried b/c it's a relatively small change. The reason why I'm keeping it in there now is that it's in testnet and we aren't 100% sure that the alternative will be shipped? So this is a de-risking move.

File: crates/indexer/src/models/token_models/token_activities.rs [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    pub transaction_version: i64,
+    pub event_account_address: String,
+    pub event_creation_number: i64,
+    pub event_sequence_number: i64,
+    pub token_data_id_hash: String,
+    pub property_version: BigDecimal,
+    pub creator_address: String,
+    pub collection_name: String,
+    pub name: String,
+    pub transfer_type: String,
```
Comment (areshand) [inline crates/indexer/src/models/token_models/token_activities.rs:33]: why do we have transfer_type in general token activity table?

Comment (areshand) [reply crates/indexer/src/models/token_models/token_activities.rs:33]: should it be more like` activity name`, ` activity_parameters`?downstream apps can parse it based on the activity name.

File: crates/indexer/src/models/token_models/token_claims.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    pub token_data_id_hash: String,
+    pub property_version: BigDecimal,
+    pub from_address: String,
+    pub to_address: String,
+    pub collection_data_id_hash: String,
+    pub creator_address: String,
+    pub collection_name: String,
+    pub name: String,
+    pub amount: BigDecimal,
+    pub table_handle: String,
```
Comment (areshand) [inline crates/indexer/src/models/token_models/token_claims.rs:31]: what is the usage of the table_handle?

File: crates/indexer/src/models/token_models/token_activities.rs (hunk 4 above)
Comment (bowenyang007) [reply crates/indexer/src/models/token_models/token_activities.rs:33]: what's activity_parameters? For ease of use we don't want to have json blobs so structured data is the way to go even though we lose some flexibility. In that case activity_name or transfer_type is just semantics. I find activity name to be a bit too literal. Transfer type is not exactly accurate but is more in line with something I'd see in an external API.

File: crates/indexer/src/models/token_models/token_claims.rs (hunk 5 above)
Comment (bowenyang007) [reply crates/indexer/src/models/token_models/token_claims.rs:31]: That's a good question, it'll be mostly for debugging. To be clear we don't have to expose all the fields in this table. I don't think that we should expose this field either.

File: crates/indexer/src/models/token_models/token_utils.rs [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+ * This file defines deserialized move types as defined in our 0x3 contracts.
+ */
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Table {
+    pub handle: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct TokenDataIdType {
```
Comment (areshand) [inline crates/indexer/src/models/token_models/token_utils.rs:24]: how about move these struct definition in a token_mod.rs?

File: crates/indexer/src/models/token_models/token_utils.rs [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+    pub creator: String,
+    collection: String,
+    name: String,
+}
+
+impl TokenDataIdType {
```
Comment (areshand) [inline crates/indexer/src/models/token_models/token_utils.rs:30]: nit: Just name it TokenDataId so it can be reused more broadly

**Pull request merged**